rbx_types/src/attributes/mod.rs

```diff
@@ -0,0 +1,259 @@
+    convert::TryFrom,
+    hash::Hash,
+    io::{self, Read, Write},
+    iter::FromIterator,
+    string::FromUtf8Error,
+};
+
+use crate::{
+    error::Error,
+    variant::{self, VariantType},
+};
+
+use thiserror::Error;
+
+use variant::Variant;
+
+mod reader;
+mod writer;
+
+use reader::get_attributes;
+
+use writer::attributes_from_map;
+
+macro_rules! create_attribute_type {
+    ({
+        $(
+            $key:ident = $number:literal,
+        )+
+    }) => {
+        #[derive(Clone, Copy, Debug, Eq, PartialEq)]
+        pub(crate) enum AttributeType {
+            $(
+                $key = $number,
+            )+
+        }
+
+        impl TryFrom<VariantType> for AttributeType {
+            type Error = AttributeError;
+
+            fn try_from(variant_type: VariantType) -> Result<Self, Self::Error> {
+                match variant_type {
+                    $(
+                        VariantType::$key => Ok(AttributeType::$key),
+                    )+
+
+                    _ => Err(AttributeError::InvalidVariantType),
+                }
+            }
+        }
+
+        impl TryFrom<u8> for AttributeType {
+            type Error = AttributeError;
+
+            fn try_from(byte: u8) -> Result<Self, Self::Error> {
+                match byte {
+                    $(
+                        $number => Ok(Self::$key),
+                    )+
+
+                    other => Err(AttributeError::InvalidValueType(other))
+                }
+            }
+        }
+    };
+}
+
+create_attribute_type!({
+    // ??? = 0x01,
+    BinaryString = 0x02,
+    Bool = 0x03,
+    // ??? = 0x04,
+    Float32 = 0x05,
+    Float64 = 0x06,
+    // ??? = 0x07,
+    // ??? = 0x08,
+    UDim = 0x09,
+    UDim2 = 0x0A,
+    // ??? = 0x0B,
+    // ??? = 0x0C,
+    // ??? = 0x0D,
+    BrickColor = 0x0E,
+    Color3 = 0x0F,
+    Vector2 = 0x10,
+    Vector3 = 0x11,
+    // ??? = 0x12,
+    // ??? = 0x13,
+    // ??? = 0x14,
+    // ??? = 0x15,
+    // ??? = 0x16,
+    NumberSequence = 0x17,
+    // ??? = 0x18,
+    ColorSequence = 0x19,
+    // ??? = 0x1A,
+    NumberRange = 0x1B,
+    Rect = 0x1C,
+});
+
+#[derive(Debug, Default, Clone, PartialEq)]
+pub struct Attributes {
+    data: HashMap<String, Variant>,
+}
+
+impl Attributes {
+    /// Creates an empty `Attributes` struct
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Reads from a serialized attributes string, and produces a new `Attributes` from it.
+    pub fn from_reader<R: Read>(reader: R) -> Result<Self, Error> {
+        Ok(Attributes {
+            data: get_attributes(reader)?,
+        })
+    }
+
+    /// Writes the attributes as a serialized string to the writer.
+    pub fn to_writer<W: Write>(&self, mut writer: W) -> Result<(), Error> {
+        let bytes = attributes_from_map(self.data.iter())?;
+        writer
+            .write_all(&bytes)
+            .map_err(AttributeError::ToWriterFail)
+            .map_err(Into::into)
```

Comment:
Can we refactor `attributes_from_map` to accept a writer instead of buffering here? It kind of defeats the point of streaming output if we aren't streaming it!